Angular: Overhaul preview renderer

ThibaudAV opened the pull request:
Issue: #11614

Work in progress, TODO list:
 - [x] fix ` Custom/Feature Module as Context` story
 - [x] add more test for Service

I did not browse all issues. but it fixes, for instance, the color background which does not change on this issue: 
 https://github.com/storybookjs/storybook/issues/11614

## What I did

### Two main goals: 
#### 1. make operation of angular rendering more clear 
(currently, all the functions are tangled in [helpers file](https://github.com/storybookjs/storybook/blob/next/app/angular/src/client/preview/angular/helpers.ts) or  ) and does not make the code very clear to understand how it works

#### 2. use a new way to add story component|template in the Angular application

Currently the component | template of the story is added dynamically with `componentFactoryResolver`. This way does not add the component in the same way as if it was used in a template (the component will not have the same life cycle) which forces Storybook to complete the missing life cycle. (see [app.component](https://github.com/storybookjs/storybook/blob/next/app/angular/src/client/preview/angular/components/app.component.ts) `ngOnChanges`, `writeValue`, `registerOnChange`, missing `registerOnTouched`, ext..)

This new approach no longer uses dynamic components. 
- For template: it's added in a component that will be used as a bootstrap component directly. All story properties are added as properties of the new wrapper component. (see this [function](https://github.com/storybookjs/storybook/pull/13215/commits/fc1c73f0906efb008cc2fd036ea92cce4ac2bbbc#diff-926bd30e1fc3b0ce350855e9c5833bbeb2e65b9ceb34b500d00a6728ad9fa4cf))
- For component: a template is generated representing the typical use of this component in a real template. This allows Angular to integrate it completely in its life cycle. (see this [function](https://github.com/storybookjs/storybook/pull/13215/commits/fc1c73f0906efb008cc2fd036ea92cce4ac2bbbc#diff-41d23c484198ea13e7a70bc6025655d05daacefa3b8c6afd58976de5288426ed))

So in this way Storybook should better simulate the real use of the story in the angular application.

### Tests 
I've checked all the story examples for angular (don't hesitate to do the same again, I think)
I found only one case that is problematic and that creates breaking changes

For the story `Component with default providers`  and `Component with overridden provider` inside [Custom/Feature Module as Context](https://github.com/storybookjs/storybook/blob/next/examples/angular-cli/src/stories/module-context/module-context.stories.ts) the `ChipComponent` cannot be added because it is not exported by ChipsModule and it is already declared by ChipsModule. So the virtual Storybook module cannot use it. :/ 
Before that, it was not a problem to dynamically add a non-exported component. Now, it's not possible anymore.

It is always possible to add a component already declared in a module because Storybook detects it and will not re-declare it. 

### Notes
~~Another point is on [NO_ERRORS_SCHEMA](https://github.com/storybookjs/storybook/pull/13215/files#diff-b30d2f4d503d54c2c1c652719030527720dafa24369fae57a34d34b779e362b8R105), it would be really nice if addon doc + controls only add the Input and Output properties of a component, and so not include private properties anymore.~~

~~Removing this will allow Storybook to return an error if component inputs & outputs no longer match story inputs & outputs. Currently, there are no errors (if I am not wrong) but I think it would be nice if Storybook returns an error in this case. This will keep the story and associated components consistent.
With my current work, it would be very handy 🙈~~

I think the problem with this issue https://github.com/storybookjs/storybook/issues/11613 is the same
The doc+controls addon returns the value of the "private" property value to in story `props`.

On my 2nd commit, I don't know how to do otherwise without editing jest configuration to run a test with `jest-preset-angular` 🤔 

## How to test

- Is this testable with Jest or Chromatic screenshots?
> yes
- Does this need a new example in the kitchen sink apps?
> 🤷‍♂️  maybe add this ? https://github.com/storybookjs/storybook/commit/0670442f3394305c81bc33199c7744e57e87dcaf
- Does this need an update to the documentation?
> 🤷‍♂️

👋  Don't hesitate to give me feedback, if you think it's not interesting. If the code is unclear or if there are too many comments or missing comments, ty

p.s : sorry for my bad english 🙈

shilman commented:
@ThibaudAV can you explain the goals of the rewrite?

ThibaudAV commented:
yes @shilman I wanted to check with CI . but I plan to do it azap

shilman commented:
cc @kroeder @Marklb

ThibaudAV committed:
feat: add enableProdMode() in RenderNgAppService when NODE_ENV !== 'development'

Marklb commented:
This looks like what I have been considering, so I will be going through it later(hopefully this afternoon) and seeing how it works.

I have only looked at the code so far, but I will list questions/suggestions I have so far.

- I understand wanting to only have the inputs/outputs in props, but I don't see that as a possibility, because the props are not specific to the component. Some of my stories use props that go with multiple or none of the components, because the story is either showing how multiple components are used together or the component itself is rendered by my code, like my stories showing how breadcrumbs work and only specify `<router-outlet></router-outlet>` as the template.
	+ Would it be possible to get the component's factory to filter it's inputs/outputs? Without Ivy, the component may need to be an `entryComponent`, but that shouldn't be an issue much longer I hope. I will check this when I get time to play with the code, but I think it would still be possible with this implementation.

- Do you think this would have any effect on the aliased inputs/outputs? I don't think it helps, but would probably prevent unintentionally overwriting variables that aren't an input/output, since they are being handled by the template. I explained my issues trying to solve the problem here: https://github.com/storybookjs/storybook/issues/12854#issuecomment-717719473

- This is what I was proposing as a way to support directives in stories. Do you think it would be reasonable to add that support to this? Related issue: #12946.

- As for `ng-content`, with the existing implementation, I was proposing `projectableNodes` as a possible solution, but if the template string creation was it's own utility function, could we iterate an array of components and use `@ViewChildren`, instead of `@ViewChild`, to still support component content with this implementation? Related issue: #10272

- You are using a `BehaviorSubject` instead of the `ReplaySubject`, but I assume it still has the `NgZone` problem. I will confirm this later, but I have a PR to hopefully fix that and you should be able to add/merge it into your implementation easily, if that solution is merged. If you want to try it out, I think it is ready to merge, but would like someone else to look it over.

- Do you think we could improve Storybook's decorator support for Angular? Related issue: #12749

Marklb commented on app/angular/src/client/preview/angular/app.token.ts:
```diff
@@ -1,4 +1,10 @@
+import { ICollection } from '../types';
 
-export const STORY = new InjectionToken<StoryFnAngularReturnType>('story');
+export const STORY_PROPS = new InjectionToken<Subject<ICollection | undefined>>('STORY_PROPS');
```
I am not completely against this, but I am wondering if it is worth it to loose the rest of the story data? It will break 1 or 2 of my utilities, but they aren't very important and I can probably fix them by getting the data of the selected story from Storybook's api when the observable emits. I am just trying to think if it could break or limit something important, which it may not.

ThibaudAV replied:
It's a private provider we shouldn't depend on it too much, I think
But we can leave the old one if we need it ? 
This one only needs props I think for the moment. but we can also use the old one (but I find it less clean ^^ )

--
For info : 
The old token type is not good ^^ this should be a `InjectionToken<Subject <StoryFnAngularReturnType>>`
--

gaetanmaisse replied:
If needed we can keep `STORY` for now and add the more specific `STORY_PROPS`.
I just checked, `STORY` wasn't part of the public API so there should be not a lot of developers using it outside the SB contributor world.

ThibaudAV commented:
I have corrected some mistakes I made while running the test 😅 

>I understand wanting to only have the inputs/outputs in props, but I don't see that as a possibility, because the props are not specific to the component. Some of my stories use props that go with multiple or none of the components, because the story is either showing how multiple components are used together or the component itself is rendered by my code, like my stories showing how breadcrumbs work and only specify <router-outlet></router-outlet> as the template. 

I speak only in component case. In template case there is no problem.
For this issue I would think more that it should be a configuration of concerned story. By default, there is an error if the `props` does not respect component input/output contract. And if this is the wish of the story, just add the shemas `NO_ERRORS_SCHEMA` or another boolean.  But don't let this be the default case 🤷‍♂️ 

But I let you choose :) I probably don't have all the context 


>Would it be possible to get the component's factory to filter it's inputs/outputs? Without Ivy, the component may need to be an entryComponent, but that shouldn't be an issue much longer I hope. I will check this when I get time to play with the code, but I think it would still be possible with this implementation.

haha I thought about it, but I think it would be a loss in performance. Load the component twice to be able to filter on properties :/ 


> Do you think this would have any effect on the aliased inputs/outputs? I don't think it helps, but would probably prevent unintentionally overwriting variables that aren't an input/output, since they are being handled by the template. I explained my issues trying to solve the problem here: #12854 (comment)

Completely. As we get as close as possible to a real use of a component (I hope) this should solve this problem in part. 
There is still the problem, for me, of the doc that automatically adds the local properties (which are not @Input/ @Output ) of the component
If the `doc` addon is smart enough to find the alias and not add the local property it would be great!   
(If I understand correctly)


> This is what I was proposing as a way to support directives in stories. Do you think it would be reasonable to add that support to this? Related issue: #12946.

With this solution it works without any problems. If my quick test is correct

> As for ng-content, with the existing implementation, I was proposing projectableNodes as a possible solution, but if the template string creation was it's own utility function, could we iterate an array of components and use @ViewChildren, instead of @ViewChild, to still support component content with this implementation? Related issue: #10272

Not sure to understand exactly 🤔  
But if we are talking about the combined use of component + template to add template in the `ng-content`? I think it's possible. I will propose to do it in another PR if this one if validated? This solution would even facilitate this feature, I think.



>You are using a BehaviorSubject instead of the ReplaySubject, but I assume it still has the NgZone problem. I will confirm this later, but I have a PR to hopefully fix that and you should be able to add/merge it into your implementation easily, if that solution is merged. If you want to try it out, I think it is ready to merge, but would like someone else to look it over.

What's the problem? we can use another observable one. it's just that this one made more sense to me for this situation. 

>Do you think we could improve Storybook's decorator support for Angular? Related issue: #12749

If I understand correctly. Yes, I don't think this solution closes doors for these features. On the contrary, I think it could make these improvements easier. What do you think about it ? 
I can see something like that 🤔 
```js
export default {
  component: YourComponent,
  decorators: [
   templateDecorator((story) => `<div style="margin: 3em">${story}</div>`)
  ],
} as Meta;
```

Marklb commented:
>I speak only in component case. In template case there is no problem.
For this issue I would think more that it should be a configuration of concerned story. By default, there is an error if the props does not respect component input/output contract. And if this is the wish of the story, just add the shemas NO_ERRORS_SCHEMA or another boolean. But don't let this be the default case 🤷‍♂️

I could be over thinking this and may need additional feedback from other to see what would be expected. It may be configurable globally using the `moduleMetadata` decorator in `preview.ts`, which would make it simple for the user to decide their default. One situation I can think of that could cause additional props could be a knob/control added to toggle a value in a decorator. As long as it is obvious that `NO_ERRORS_SCHEMA` is needed to fix that, then it's no problem. Would user's that don't often write tests or use a library that handles the boilerplate know that `NO_ERRORS_SCHEMA` exists? Would we be able to determine if that should be suggested without checking the inputs/outputs on the factory?

> haha I thought about it, but I think it would be a loss in performance. Load the component twice to be able to filter on properties :/

> Completely. As we get as close as possible to a real use of a component (I hope) this should solve this problem in part.
> There is still the problem, for me, of the doc that automatically adds the local properties (which are not @Input/ @output ) of the component
> If the `doc` addon is smart enough to find the alias and not add the local property it would be great!
> (If I understand correctly)

It's been a while since I looked at what `resolveComponentFactory` does, but I thought it was basically just a lookup for the factory that creates the component. The inputs/outputs are on the factory object, so the component doesn't need to be created. I would need to look into it more to see what the impact would be, but I am not sure if there is another way to reliably get the `templateName` of inputs/outputs currently.

> What's the problem? we can use another observable one. it's just that this one made more sense to me for this situation.

I forgot to reference the PR: #12382
The problem that I am fixing with that issue is the observable not always emitting in the zone, since an event triggered by anything outside our story, such as addons, are not hooked by NgZone. I just tried a story on your branch and it does have the problem. As long as that PR doesn't have a problem I am not seeing, it is a simple fix.

ThibaudAV commented:
> I forgot to reference the PR: #12382
The problem that I am fixing with that issue is the observable not always emitting in the zone, since an event triggered by anything outside our story, such as addons, are not hooked by NgZone. I just tried a story on your branch and it does have the problem. As long as that PR doesn't have a problem I am not seeing, it is a simple fix.

I propose you an alternative solution in PR #12382 
Whatever the choice I think that both are compatible with this pr no ?

ThibaudAV commented:
For input/output recovery, I can propose you this solution which doesn't require the angular engine and is based on Typescript. I don't think the Input/Output convention will change for a long time. what do you think about it ? 
```js
      const getComponentIOPropertiesName = (
        compType: any
      ): { inputs: string[]; outputs: string[] } | undefined => {
        const decoratorKey = '__prop__metadata__';
        const propsDecorators: Record<string, any[]> =
          Reflect &&
          Reflect.getOwnPropertyDescriptor &&
          Reflect.getOwnPropertyDescriptor(compType, decoratorKey)
            ? Reflect.getOwnPropertyDescriptor(compType, decoratorKey).value
            : compType[decoratorKey];

        if (!propsDecorators) {
          return undefined;
        }

        return Object.entries(propsDecorators).reduce(
          (previousValue, [key, value]) => {
            if (value[0].ngMetadataName === 'Input') {
              const propertyName = value[0].bindingPropertyName ?? key;
              return {
                ...previousValue,
                inputs: [...previousValue.inputs, propertyName],
              };
            }
            if (value[0].ngMetadataName === 'Output') {
              const propertyName = value[0].bindingPropertyName ?? key;
              return {
                ...previousValue,
                outputs: [...previousValue.outputs, propertyName],
              };
            }
            return previousValue;
          },
          {
            inputs: [],
            outputs: [],
          }
        );
      };
      console.log(getComponentIOPropertiesName(storyObj.component));
```
if it suits you. We will have to define what to do with it because I'm not sure I understand all your concerns  🙃

Marklb commented:
When I mentioned filtering, I was just meaning to only select the props that are an input or output, to avoid the `NO_ERRORS_SCHEMA` issue that you mentioned, but I think that the inputs/outputs have to be known either way to build the template.

That `getComponentIOPropertiesName` function would miss inputs in `@Component` and extended classes. So, in the following example, `baseInp` and `inp2` would be missed.
```ts
@Component({ selector: '' })
class BaseExample {
  @Input() baseInp1: string
}

@Component({
  selector: 'story-example',
  template: ``,
  inputs: [ 'inp2' ]
})
class StoryExample extends BaseExample {
  @Input() inp1: string
  @Input('reInp3') inp3: string
}
```

Marklb commented on app/angular/src/client/preview/angular/ComponentClassFromStoryComponent.ts:
```diff
@@ -0,0 +1,115 @@
+  return ngComponentDecorator;
+};
+
+const toInputsOutputs = (props: ICollection = {}) => {
+  return Object.entries(props).reduce(
+    (previousValue, [key, value]) => {
+      if (typeof value === 'function') {
+        return { ...previousValue, outputs: { ...previousValue.outputs, [key]: value } };
+      }
+
+      return { ...previousValue, inputs: { ...previousValue.inputs, [key]: value } };
+    },
+    { inputs: {}, outputs: {} } as { inputs: Record<string, any>; outputs: Record<string, any> }
+  );
+};
```
`typeof value === 'function'` is not enough to determine an output.

The following story will incorrectly think the input is an output:
```ts
@Component({
  selector: 'story-example',
  template: `[{{ filter ? filter(list) : list }}]`
})
class StoryExample {
  @Input() filter: (list: string[]) => string[]

  list: string[] = [ 'a', 'b', 'c' ]
}

export const Example = () => ({
  props: {
    filter: (list: string[]) => list.filter(x => x !== 'b')
  },
  component: StoryExample
})
```

ThibaudAV replied:
if we have the right list of Inputs/Outputs from the component
we could make the distinction with it. 

And for all the props not found, it is that they are local to the component. We could also add them to the component in the ngAfterViewInit only once at the init.
This would keep the current behavior that allows to modify these properties. 🤷‍♂️ 

I'm wrong? if you're ok with it I can commit with this behavior.

ThibaudAV commented:
@Marklb Do you think this is the proper approach to use? 
If yes I can add a commit with tests for all cases. To find all inputs/outputs

Marklb commented:
> @Marklb Do you think this is the proper approach to use? 
> If yes I can add a commit with tests for all cases. To find all inputs/outputs

Yes. At first I was leaning toward just using the factory, but now I am liking your approach more. One reason is because it could potentially be used by other features that aren't running in Angular, like #12438

ThibaudAV committed:
feat: add function to find Inputs & Outputs in an Angular component

The test allows to verify the function by using angular engine

ThibaudAV committed:
refactor: rework story's props association with the component's properties

- allow story with component to override local properties
- handles Input with `bindingPropertyName`

ThibaudAV commented:
@Marklb I've added 2 commits on the last discussed subjects

I have also added more "complete" tests. In order to be able to be more safe about the link between the component and the props.

Tell me what you think? Do you think this pr is a good thing ? and that we should finish it and test it well, to merge it in `next` ?

Don't hesitate if you want to add code or modify the existing one.

gaetanmaisse commented on app/angular/src/client/preview/angular/ComponentClassFromStoryComponent.ts:
```diff
@@ -0,0 +1,146 @@
+    .map((i) => i.templateName);
+  const outputs = ngComponentInputsOutputs.outputs
+    .filter((o) => o.templateName in props)
+    .map((i) => i.templateName);
```
`i` should be `o` 🙈

gaetanmaisse commented on app/angular/src/client/preview/angular/ComponentClassFromStoryComponent.ts:
```diff
@@ -0,0 +1,146 @@
+} from './NgComponentAnalyzer';
+import { RenderNgAppService } from './RenderNgAppService';
+
+const distinctComponentInputsOutputsWithProps = (
```
If I understand correctly the function I would suggest renaming it to something like `getNamesOfInputsOutputsDefinedInProps`

gaetanmaisse commented on app/angular/src/client/preview/angular/ComponentClassFromStoryComponent.ts:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    ngAfterViewInit(): void {
+      // Initializes properties that are not Inputs | Outputs
+      // Allows story props to override local component properties
+      initialOtherProps.forEach((p) => {
+        (this.storyComponentElementRef as any)[p] = initialProps[p];
+      });
```
If I correctly followed all your comments in this PR (and in another issue), you're suggesting to only assign `initialInputs` and `initialOutputs`, and not the whole `initialProps` which can lead to setting values to "private/internal" values? 

In my opinion, it makes perfect sense and should work like that to better reflect the real-life use of the component. However, it's a breaking change so we can not introduce such behavior until 7.0.

ThibaudAV replied:
In this PR for the moment it works as before. 
But yes, I suggest that this is no longer the default behavior. 
I think, it can be the subject of a dedicated PR or we could first add a warning message.
Then add a flag (like `overrideInternalComponentProperty`) to make it optional. 🤷‍♂️

gaetanmaisse commented on app/angular/src/client/preview/angular/NgComponentAnalyzer.test.ts:
```diff
@@ -0,0 +1,116 @@
+
+import { getComponentInputsOutputs } from './NgComponentAnalyzer';
+
+describe('getComponentInputsOutputs', () => {
```
Awesome test suite, it will make us much more confident about everything related to component properties detection 💪🏻

gaetanmaisse commented on app/angular/src/client/preview/angular/NgComponentAnalyzer.ts:
```diff
@@ -0,0 +1,106 @@
+ * Returns all component decorator properties
+ * is used to get all `@Input` and `@Output` Decorator
+ */
+export const getComponentPropsDecoratorMetadata = (component: any) => {
+  const decoratorKey = '__prop__metadata__';
+  const propsDecorators: Record<string, (Input | Output)[]> =
+    Reflect &&
+    Reflect.getOwnPropertyDescriptor &&
+    Reflect.getOwnPropertyDescriptor(component, decoratorKey)
+      ? Reflect.getOwnPropertyDescriptor(component, decoratorKey).value
+      : component[decoratorKey];
+
+  return propsDecorators || undefined;
```
I trust you 🙈

ThibaudAV replied:
This undefined is not really necessary 
it's in case propsDecorators returns null 🤷‍♂️

gaetanmaisse commented on app/angular/src/client/preview/angular/NgModulesAnalyzer.ts:
```diff
@@ -0,0 +1,59 @@
+ *
+ * Checks recursively if the component has already been declared in all import Module
+ */
+export const getExistenceOfComponentInModules = (
```
`isComponentAlreadyDeclaredInModules`?

ThibaudAV replied:
haha I had kept the original name of this function. Good suggestion I change it ;)

gaetanmaisse commented:
I really like the separation between the "component factories" for stories with `template` and the ones with `component`.
Plus, we are no more relying on dynamically generated components, this looks to simplify lifecycle management of the components. Some part of the code isn't straight forward, but there are some good tests, so I feel really confident.

@ThibaudAV if you want we can pair to find out why tests aren't working on CI + fix the only broken story?

I think it's a great first step that sets good bases for adding features and fix bugs more easily in 6.2.

@Marklb do you have in mind other things that should land in this PR?

ThibaudAV commented:
> @ThibaudAV if you want we can pair to find out why tests aren't working on CI + fix the only broken story?
@gaetanmaisse the problem seems to be fixed with a small dependency update

I added an option `useLegacyRendering` which allows to use the old renderer in case of default on a story

ThibaudAV committed:
fix: angluar-cli export Chip component to be accessible by storybook render module

Marklb commented:
> @Marklb do you have in mind other things that should land in this PR?

@gaetanmaisse Anything else I can think of at the moment would be new features made possible/easier from this. So, they would be separate PR's.

shilman commented:
@Marklb @gaetanmaisse sounds like this PR is good to go then? @ThibaudAV Can you get the tests to pass?

shilman commented on app/angular/src/client/preview/types.ts:
```diff
@@ -32,4 +32,6 @@ export interface StoryFnAngularReturnType {
   template?: string;
   styles?: string[];
+  /** Uses legacy angular rendering engine that use dynamic component */
+  useLegacyRendering?: boolean;
```
WDYT about making this a framework option. User would write the following in `.storybook/main.js`:

```js
module.exports = {
  angularOptions: {
    legacyRendering: true,
  }
}
```

Another option would be to use story parameters if we want to control this on a per story basis. To configure globally, `.storybook/preview.js`:

```js
export const parameters = {
  angularLegacyRendering: true,
}
```

Both of these would be more "storybook-like" than adding a new field to the story function return type.

Marklb replied:
I like the idea of the framework option, because that projects rendering would be consistent, instead of potentially introducing bugs from the previous renderer not getting cleaned up.

As long as the renderer's clean themselves up when switching, then users with a large project may have a reason to migrate a few stories at a time. So, a parameter may be the better choice to avoid users not being able to start migrating, because of a bug or situation we didn't consider.

shilman reviewed (changes requested):
Small suggestion about configuration via framework options or parameters. See comment line @ThibaudAV @gaetanmaisse

ThibaudAV committed:
test: add test for RenderNgAppService

Configure app/angular package as jest projects
Allow to use jest-preset-angular for app/angular tests

ThibaudAV commented:
@shilman I added it as a parameters :
https://github.com/storybookjs/storybook/pull/13215/commits/30d3a1916edbc317fda362636e84c555c7a27209#diff-2301c93915662308821a28d0c92f96d5bb1253c16872c6f3aeac982087ddc006R31
I hope this is the goot way to do it ? 

I also corrected tests

ThibaudAV committed:
feat: add a rewrite of client angular rendering

A `useLegacyRendering` flag allows to use old rendering engine

shilman committed:
MIGRATION.md for new angular renderer

kroeder commented:
Impressive PR, great job! 💪

shilman commented:
cc @mandarini
